addons/hr_holidays/models/hr_leave.py

```diff
@@ -359,21 +352,32 @@ def _compute_state(self):
         for leave in self:
             leave.state = 'confirm' if leave.validation_type != 'no_validation' else 'draft'
 
+    @api.depends('holiday_type', 'employee_id', 'department_id', 'mode_company_id')
+    def _compute_resource_calendar_id(self):
+        for leave in self:
+            calendar = False
+            if leave.holiday_type == 'employee':
+                calendar = leave.employee_id.resource_calendar_id
+            elif leave.holiday_type == 'department':
+                calendar = leave.department_id.company_id.resource_calendar_id
+            elif leave.holiday_type == 'company':
+                calendar = leave.mode_company_id.resource_calendar_id
+
+            leave.resource_calendar_id = calendar or self.env.company.resource_calendar_id
+
     @api.depends('request_date_from_period', 'request_hour_from', 'request_hour_to', 'request_date_from', 'request_date_to',
                  'request_unit_half', 'request_unit_hours', 'employee_id')
     def _compute_date_from_to(self):
         for holiday in self:
-            if holiday.request_date_from and holiday.request_date_to and holiday.request_date_from > holiday.request_date_to:
-                holiday.request_date_to = holiday.request_date_from
             if not holiday.request_date_from:
                 holiday.date_from = False
             elif not holiday.request_unit_half and not holiday.request_unit_hours and not holiday.request_date_to:
                 holiday.date_to = False
             else:
-                if holiday.request_unit_half or holiday.request_unit_hours:
+                if (holiday.request_unit_half or holiday.request_unit_hours) and not holiday.request_date_to == holiday.request_date_from:
```

Comment:
It would generate an extra write, which is not great for performance and it also makes debugging harder. I do agree it's a little long but I don't think it's that big of a deal, or do you feel very strongly about this?